garbagecollector: use contextual logging

ncdc opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Switch the garbage collector code to use contextual and structured logging.


#### Which issue(s) this PR fixes:

Part of kubernetes/enhancements#3077

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Migrated the Kubernetes object garbage collector (within `kube-controller-manager`) to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/sig api-machinery
/wg structured-logging
/assign @pohly 
/cc @pohly

ncdc commented:
/cc @stevekuznetsov @nrb @astefanutti

mengjiao-liu commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -397,16 +412,27 @@ func (gc *GarbageCollector) attemptToDeleteWorker(ctx context.Context, item inte
 func (gc *GarbageCollector) isDangling(ctx context.Context, reference metav1.OwnerReference, item *node) (
 	dangling bool, owner *metav1.PartialObjectMetadata, err error) {
 
+	logger := klog.FromContext(ctx)
+	logger = klog.LoggerWithValues(logger,
+		"object.uid", item.identity.UID,
+		"ownerReference.apiVersion", reference.APIVersion,
+		"ownerReference.kind", reference.Kind,
+		"ownerReference.name", reference.Name,
+	)
+
```
However, if ContextualLogging feature gate is not enabled, these keyValues pairs are lost and cannot be rolled back to default.

/cc @pohly

stevekuznetsov replied:
Is that true? Without contextual logging enabled, `klog.FromContext()` will return a default sentinel value, but it's still a structural logger. This PR does not use `klog.NewContext()` to pass information through the logger in the context, and *that* would be the only bit that doesn't work without the feature flag, right?

ncdc replied:
```
func LoggerWithValues(logger Logger, kv ...interface{}) Logger {
	if contextualLoggingEnabled {
		return logger.WithValues(kv...)
	}
	return logger
}
```

😢

pohly replied:
Yes, we had to make these functions a NOP to satisfy the feature gate criteria for alpha features. For these conversions that means that we can start to use some of the new features to make logs better, but it won't have an effect unless the feature gets enabled.

That also implies that the values still need to be logged below. They will only get logged once when the feature is enabled.

stevekuznetsov replied:
@pohly this is pretty surprising to me - doesn't it mean that you can't use any contextual logging features whatsoever until the feature gates default to on... ?

pohly replied:
You can enable the feature gate via the usual `--feature-gates ContextualLogging=true`.

stevekuznetsov replied:
Right, but I'm asking about the broader case - if we make a change to some part of the k8s code-base to add context to a logger and remove it from a structured log message, it will regress logging output for the default feature gate case, no?

stevekuznetsov replied:
Or I guess @pohly what's the policy for "regression" on logging?

pohly replied:
> if we make a change to some part of the k8s code-base to add context to a logger and remove it from a structured log message

We can't do that in a single step. We first have to add contextual logging, then at a later point when it is considered "always on" (= beta or GA) the then redundant key/value pairs in the `logger.Info` or `logger.Error` can be removed.

The implementation intentionally filters out redundant key/value pairs when contextual logging is on and WithValues was used, to support such a transition period.

I know that it is annoying, but we had to support turning off the feature.

pohly replied:
> what's the policy for "regression" on logging?

Log messages can change at any time, there are no stability guarantees. But we obviously don't want to make logs less useful because of the migration to contextual logging, so values will have to be in both (`WithValues` and parameters) for a while.

ncdc replied:
I just pushed an update that preserves

ncdc replied:
Looks like I might have missed some - checking

stevekuznetsov commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -287,20 +296,22 @@ func (gb *GraphBuilder) IsSynced() bool {
-	defer klog.Infof("GraphBuilder stopping")
+func (gb *GraphBuilder) Run(ctx context.Context) {
+	logger := klog.FromContext(ctx)
+	logger = klog.LoggerWithValues(logger, "component", "GraphBuilder")
```
*do* we want to add this to the context so that the downstream calls have it?

ncdc replied:
Probably

pohly replied:
Let's add a name in app/core.go.

pohly replied:
No need to change anything regarding this in the PR, I'll cover it in a final cleanup PR that'll make all controllers consistent.

fedebongio commented:
/triage accepted

shivanshu1333 commented:
/assign

pohly commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -514,7 +514,7 @@ func startGarbageCollectorController(ctx context.Context, controllerContext Cont
 	// Periodically refresh the RESTMapper with new discovery information and sync
 	// the garbage collector.
-	go garbageCollector.Sync(discoveryClient, 30*time.Second, ctx.Done())
+	go garbageCollector.Sync(ctx, discoveryClient, 30*time.Second)
```
Let's add a name here, see https://github.com/kubernetes/kubernetes/pull/113916#discussion_r1025594468

pohly commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -204,26 +210,30 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.ServerResourcesInterf
 				newResources = GetDeletableResources(discoveryClient)
 				if len(newResources) == 0 {
-					klog.V(2).Infof("no resources reported by discovery (attempt %d)", attempt)
+					logger.V(2).Info("no resources reported by discovery", "attempt", attempt)
```
Strictly speaking, this would need to look up the logger again because it runs with a different `ctx` than the surrounding function. But in practice it probably doesn't matter, so let's leave it.

pohly commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -397,16 +412,27 @@ func (gc *GarbageCollector) attemptToDeleteWorker(ctx context.Context, item inte
 
+	logger := klog.FromContext(ctx)
+	logger = klog.LoggerWithValues(logger,
+		"object.uid", item.identity.UID,
```
I believe the guideline is to not use a dot in the key name.

pohly replied:
I would use `objectUID`.

sftim replied:
Would `uid` work? If not, why not?

pohly replied:
Just `uid` is a bit vague. If any log call then logs another `uid`, that value will get logged instead of the `item.identity.UID` value.

pohly replied:
@ncdc: can you replaced `object.uid` with `objectUID`?

Instead of logging each individual ownerReference field, you can use `"owner", reference`.

Please also add the change note as mentioned in https://github.com/kubernetes/kubernetes/pull/113471#issuecomment-1367556909

ncdc replied:
Updated and added release note.

sftim commented:
Suggestion for the release note:
```markdown
Migrated the Kubernetes object garbage collector (within `kube-controller-manager`) to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

AIUI, end users could observe this change.

ncdc committed:
garbagecollector: use contextual logging

Signed-off-by: Andy Goldstein <andy.goldstein@redhat.com>

ncdc commented:
/reopen

Something is maybe weird with GitHub not seeing my force push

ncdc commented:
/reopen

ncdc commented:
Ok, got GH to pick up my rebase.

@pohly updated, PTAL! Thx 😄

pohly commented:
I haven't checked whether that hit you, but one of the weird things with GitHub is that you can force-push while a PR is open, but you cannot reopen after a force-push.

pohly commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -610,30 +715,41 @@ func (gc *GarbageCollector) attemptToDeleteItem(ctx context.Context, item *node)
-		klog.V(2).Infof("remove DeleteDependents finalizer for item %s", item.identity)
-		return gc.removeFinalizer(item, metav1.FinalizerDeleteDependents)
+		logger.V(2).Info("remove DeleteDependents finalizer for item", "item", item.identity)
+		return gc.removeFinalizer(ctx, item, metav1.FinalizerDeleteDependents)
```
There's a slight bug here: `logger` was updated, but the `ctx` still uses the old logger.

I think all of these `gc` methods should better take a `logger` instead of `ctx`, then the need to create a new context with the updated logger goes away and those functions don't need to call `FromContext`.

ncdc replied:
Some of the gc methods already take in a context (e.g. for making a client call). Would you rather they become more uniform (take in a context when they didn't before), or have a mix of taking in a logger (new addition) or a context (existing code)?

pohly replied:
I think a mix is fine. It's always a case-by-case decision.

pohly commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -185,7 +190,7 @@ func (gb *GraphBuilder) controllerFor(resource schema.GroupVersionResource, kind
 // Run. Monitors are NOT started as part of the sync. To ensure all existing
 // monitors are started, call startMonitors.
-func (gb *GraphBuilder) syncMonitors(resources map[schema.GroupVersionResource]struct{}) error {
+func (gb *GraphBuilder) syncMonitors(ctx context.Context, resources map[schema.GroupVersionResource]struct{}) error {
```
Same here: the code might become simpler when passing around a `logger`.

pohly commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -287,20 +296,21 @@ func (gb *GraphBuilder) IsSynced() bool {
-func (gb *GraphBuilder) Run(stopCh <-chan struct{}) {
-	klog.Infof("GraphBuilder running")
-	defer klog.Infof("GraphBuilder stopping")
+func (gb *GraphBuilder) Run(ctx context.Context) {
```
Here `ctx` makes sense because it also covers cancellation.

pohly commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -650,14 +687,33 @@ func (gb *GraphBuilder) processGraphChanges() bool {
 				if len(observedIdentity.Namespace) > 0 && dep.identity.Namespace != observedIdentity.Namespace {
 					// Namespace mismatch, this is definitely wrong
-					klog.V(2).Infof("node %s references an owner %s but does not match namespaces", dep.identity, observedIdentity)
+					logger.V(2).Info("node references an owner but does not match namespaces", "apiVersion", event.gvk.GroupVersion().String(),
```
This logs a lot more information than before.

At first glance that seems useful, but what about performance implications?

ncdc replied:
I was mostly copying the data from the LoggerWithValues call from above (as is required in this transitional state). If you'd prefer, I can revert back so each Info statement only logs what was previously being logged?

pohly replied:
Let's log only what was being logged before.

ncdc commented:
@pohly thanks for the feedback! Just pushed a code review commit.

ncdc commented:
/test pull-kubernetes-integration

pohly commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -498,12 +543,29 @@ func ownerRefsToUIDs(refs []metav1.OwnerReference) []types.UID {
 	if item.isBeingDeleted() && !item.isDeletingDependents() {
-		klog.V(5).Infof("processing item %s returned at once, because its DeletionTimestamp is non-nil", item.identity)
+		logger.V(5).Info("processing item returned at once, because its DeletionTimestamp is non-nil",
+			"object", klog.KRef(item.identity.Namespace, item.identity.Name),
```
There are still cases where the modified log calls include more information than before. Can you take out all additional key/value pairs?

ncdc replied:
Sorry, thought I got them all. Will do on Monday!

pohly commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -342,18 +354,21 @@ func (gc *GarbageCollector) attemptToDeleteWorker(ctx context.Context, item inte
 	}
 
+	logger := klog.FromContext(ctx)
+	logger = klog.LoggerWithValues(logger, "item", n)
```
I'm having second thoughts about this and other `WithValues` calls: before we add those, we should have some kind of before/after comparison of the performance impact.

@ncdc: can we perhaps merge this PR with just the conversion of the existing calls, then enhance the log content in a second PR? Then we can do the benchmarking for that one.

ncdc replied:
@pohly I just pushed another commit that should undo the extra logged fields. Is that looking better (and did I miss anything)? Thanks!

pohly commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -151,21 +151,25 @@ func (gc *GarbageCollector) Run(ctx context.Context, workers int) {
-	klog.Infof("Starting garbage collector controller")
-	defer klog.Infof("Shutting down garbage collector controller")
+	logger := klog.FromContext(ctx)
+	logger.Info("Starting controller", "controller", "garbage-collector")
+	defer logger.Info("Shutting down controller", "controller", "garbage-collector")
```
```suggestion
	logger.Info("Starting controller", "controller", "garbagecollector")
	defer logger.Info("Shutting down controller", "controller", "garbagecollector")
```

ncdc replied:
Fixed

pohly commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -384,7 +398,7 @@ func (gc *GarbageCollector) attemptToDeleteWorker(ctx context.Context, item inte
 		// otherwise a virtual node for an item added AND removed during watch reestablishment can get stuck in the graph and never removed.
 		// see https://issue.k8s.io/56121
-		klog.V(5).Infof("item %s hasn't been observed via informer yet", n.identity)
+		logger.V(5).Info("item hasn't been observed via informer yet", "item", n.identity)
```
Should this log `n` for "item" as in the other log calls above?

ncdc replied:
I don't think so, unless you want to make a change to what is being logged here?

pohly replied:
I agree that we shouldn't log more than before. However, consistent usage of a certain key like "item" is important. We can't use it for two different things.

The inverse is also true: we shouldn't use different keys for the same thing.

Given that most of the log calls seems to log `n.identity`, let's also use that above.

ncdc replied:
Switch to item/n.identity

pohly commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -438,14 +465,20 @@ func (gc *GarbageCollector) isDangling(ctx context.Context, reference metav1.Own
-		klog.V(5).Infof("object %s's owner %s/%s, %s is not found, UID mismatch", item.identity.UID, reference.APIVersion, reference.Kind, reference.Name)
+		logger.V(5).Info("object's owner is not found, UID mismatch",
+			"objectUID", item.identity.UID,
+			"reference", reference,
```
```suggestion
			"owner", reference,
```

ncdc replied:
Fixed

pohly commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -498,12 +531,20 @@ func ownerRefsToUIDs(refs []metav1.OwnerReference) []types.UID {
 	if item.isBeingDeleted() && !item.isDeletingDependents() {
-		klog.V(5).Infof("processing item %s returned at once, because its DeletionTimestamp is non-nil", item.identity)
+		logger.V(5).Info("processing item returned at once, because its DeletionTimestamp is non-nil",
+			"item", item.identity,
```
Same as above: what needs to be logged for "item"?

ncdc replied:
Since the current code logs item.identity, we should probably keep it that way?

pohly commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -515,45 +556,63 @@ func (gc *GarbageCollector) attemptToDeleteItem(ctx context.Context, item *node)
 		// the virtual node from GraphBuilder.uidToNode.
-		klog.V(5).Infof("item %v not found, generating a virtual delete event", item.identity)
+		logger.V(5).Info("item not found, generating a virtual delete event",
+			"item", item.identity,
```
"item"?

ncdc replied:
Same as above - switching to item would be a change

pohly commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -515,45 +556,63 @@ func (gc *GarbageCollector) attemptToDeleteItem(ctx context.Context, item *node)
 	if latest.GetUID() != item.identity.UID {
-		klog.V(5).Infof("UID doesn't match, item %v not found, generating a virtual delete event", item.identity)
+		logger.V(5).Info("UID doesn't match, item not found, generating a virtual delete event",
+			"item", item.identity,
```
"item"?

ncdc replied:
Ditto

pohly commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -515,45 +556,63 @@ func (gc *GarbageCollector) attemptToDeleteItem(ctx context.Context, item *node)
 	if len(ownerReferences) == 0 {
-		klog.V(2).Infof("object %s's doesn't have an owner, continue on next item", item.identity)
+		logger.V(2).Info("object doesn't have an owner, continue on next item",
+			"item", item.identity,
```
"item"?

pohly replied:
Okay, I'll stop calling this out :sweat_smile: 

Please check this in all places.

pohly commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -610,30 +674,37 @@ func (gc *GarbageCollector) attemptToDeleteItem(ctx context.Context, item *node)
-			klog.V(2).Infof("adding %s to attemptToDelete, because its owner %s is deletingDependents", dep.identity, item.identity)
+			logger.V(2).Info("adding dependent to attemptToDelete, because its owner is deletingDependents",
+				"item", item.identity,
+				"dependent", dep.identity,
```
Also here: I don't know how large `dep` is, but perhaps it should get logged?

pohly replied:
OTOH, that makes the log larger than it used to be, something that goes beyond merely translating the calls.

ncdc replied:
Keeping dep.identity

pohly commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -572,22 +575,22 @@ func (gb *GraphBuilder) addUnblockedOwnersToDeleteQueue(removed []metav1.OwnerRe
 	}
 	if startsWaitingForDependentsDeleted(oldObj, newAccessor) {
-		klog.V(2).Infof("add %s to the attemptToDelete, because it's waiting for its dependents to be deleted", n.identity)
+		logger.V(2).Info("add node to attemptToDelete, because it's waiting for its dependents to be deleted", "node", n.identity)
```
Should this be "item" to match the other calls?

ncdc replied:
Updated

pohly commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -572,22 +575,22 @@ func (gb *GraphBuilder) addUnblockedOwnersToDeleteQueue(removed []metav1.OwnerRe
+func (gb *GraphBuilder) processTransitions(logger klog.Logger, oldObj interface{}, newAccessor metav1.Object, n *node) {
 	if startsWaitingForDependentsOrphaned(oldObj, newAccessor) {
-		klog.V(5).Infof("add %s to the attemptToOrphan", n.identity)
+		logger.V(5).Info("add node to attemptToOrphan", "node", n.identity)
```
Should this be "item" to match the other calls?

ncdc replied:
Updated

pohly commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -632,7 +635,17 @@ func (gb *GraphBuilder) processGraphChanges() bool {
+	logger.V(5).Info("GraphBuilder process object",
+		"apiVersion", event.gvk.GroupVersion().String(),
+		"kind", event.gvk.Kind,
+		"namespace", accessor.GetNamespace(),
+		"name", accessor.GetName(),
```
```suggestion
		"object", klog.KObj(accessor),
```

Does this work?

ncdc replied:
Changed

pohly commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -650,14 +663,20 @@ func (gb *GraphBuilder) processGraphChanges() bool {
 					// Namespace mismatch, this is definitely wrong
-					klog.V(2).Infof("node %s references an owner %s but does not match namespaces", dep.identity, observedIdentity)
+					logger.V(2).Info("node references an owner but does not match namespaces",
+						"node", dep.identity,
```
"item"?

ncdc replied:
Updated

pohly commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -696,10 +715,12 @@ func (gb *GraphBuilder) processGraphChanges() bool {
 		if !found {
-			klog.V(5).Infof("%v doesn't exist in the graph, this shouldn't happen", accessor.GetUID())
+			logger.V(5).Info("node doesn't exist in the graph, this shouldn't happen",
+				"node", accessor.GetUID(),
```
"node/item/UID"? Not sure...

ncdc replied:
Went with item

pohly reviewed (changes requested):
I took another, closer look at string constants and have some suggestions/questions.

ncdc commented:
Flaked on `pkg/controller/volume/attachdetach/reconciler Test_Run_OneVolumeDetachFailNodeWithReadWriteOnce`

/retest

pohly commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -610,30 +674,37 @@ func (gc *GarbageCollector) attemptToDeleteItem(ctx context.Context, item *node)
-			"objectUID", item.identity.UID, "kind", item.identity.Kind, "propagationPolicy", policy)
+		logger.V(2).Info("Deleting object",
+			"object", klog.KRef(item.identity.Namespace, item.identity.Name),
+			"objectUID", item.identity.UID,
```
Elsewhere, this would have been `"item", item`. Should this here be changed to match that or should all other lines be matched to this here ("object" + "objectUID")?

Have you checked that `"item", item.identity` leads to a useful log entry? In other words, what is the output?

If it's useful, then we should use a single key ("item") with a more complex struct (`item.identity`).

Sorry for all the back and forth on this.

ncdc replied:
I'll check and report back

ncdc replied:
`identity` is an `objectReference`, which has a `String()` implementation:

```go
return fmt.Sprintf("[%s/%s, namespace: %s, name: %s, uid: %s]", s.APIVersion, s.Kind, s.Namespace, s.Name, s.UID)
```

pohly replied:
That explains why this code here doesn't use `item.identity` as value.

But we can do better than this. A single key/value pair is clearly the right solution, we just need to ensure that it gets logged well in text *and* JSON format. Let me try something...

pohly replied:
You can add this commit to your branch and then pass a single key/value pair. Note that there are other places which log individual values, for example "GraphBuilder process object" in graph_builder.go.

```patch
From 91bef2448f6f0a88f140a49d60795b30ae7f6a88 Mon Sep 17 00:00:00 2001
From: Patrick Ohly <patrick.ohly@intel.com>
Date: Mon, 6 Mar 2023 21:37:31 +0100
Subject: [PATCH 1/1] garbagecollector: structured logging of objectReference

When using JSON as output format, we want objectReference values to be
represented as a struct. For example, "item" is such a value:

{"ts":1678135015708.349,"caller":"garbagecollector/garbagecollector.go:595","msg":"classify object references","v":5,"item":{"name":"dra-test-driver-g4tkd","namespace":"dra-1830","apiVersion":"v1","uid":"c3f88616-7282-488c-887c-3f04291e6f4f"},"solid":null,"dangling":[{"apiVersion":"apps/v1","kind":"ReplicaSet","name":"dra-test-driver","uid":"dbe9a90c-9dfd-4ad0-8395-e5fa228f9851","controller":true,"blockOwnerDeletion":true}],"waitingForDependentsDeletion":null}
---
 pkg/controller/garbagecollector/graph.go | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/pkg/controller/garbagecollector/graph.go b/pkg/controller/garbagecollector/graph.go
index b2fd8002d5c..de017b4c40b 100644
--- a/pkg/controller/garbagecollector/graph.go
+++ b/pkg/controller/garbagecollector/graph.go
@@ -20,6 +20,8 @@ import (
 	"fmt"
 	"sync"
 
+	"github.com/go-logr/logr"
+
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 )
@@ -30,10 +32,29 @@ type objectReference struct {
 	Namespace string
 }
 
+// String is used when logging an objectReference in text format.
 func (s objectReference) String() string {
 	return fmt.Sprintf("[%s/%s, namespace: %s, name: %s, uid: %s]", s.APIVersion, s.Kind, s.Namespace, s.Name, s.UID)
 }
 
+// MarshalLog is used when logging an objectReference in JSON format.
+func (s objectReference) MarshalLog() interface{} {
+	return struct {
+		Name       string    `json:"name"`
+		Namespace  string    `json:"namespace"`
+		APIVersion string    `json:"apiVersion"`
+		UID        types.UID `json:"uid"`
+	}{
+		Namespace:  s.Namespace,
+		Name:       s.Name,
+		APIVersion: s.APIVersion,
+		UID:        s.UID,
+	}
+}
+
+var _ fmt.Stringer = objectReference{}
+var _ logr.Marshaler = objectReference{}
+
 // The single-threaded GraphBuilder.processGraphChanges() is the sole writer of the
 // nodes. The multi-threaded GarbageCollector.attemptToDeleteItem() reads the nodes.
 // WARNING: node has different locks on different fields. setters and getters
-- 
2.30.2

```

ncdc replied:
Thanks for this! Just pushed with this and more logging changes to hopefully do what you're looking for 😄 . Note, I didn't touch `GraphBuilder process object` because that is the data that comes from an event, before there is a node. I think it makes sense to keep these separate.

pohly committed:
garbagecollector: structured logging of objectReference

When using JSON as output format, we want objectReference values to be
represented as a struct. For example, "item" is such a value:

{"ts":1678135015708.349,"caller":"garbagecollector/garbagecollector.go:595","msg":"classify object references","v":5,"item":{"name":"dra-test-driver-g4tkd","namespace":"dra-1830","apiVersion":"v1","uid":"c3f88616-7282-488c-887c-3f04291e6f4f"},"solid":null,"dangling":[{"apiVersion":"apps/v1","kind":"ReplicaSet","name":"dra-test-driver","uid":"dbe9a90c-9dfd-4ad0-8395-e5fa228f9851","controller":true,"blockOwnerDeletion":true}],"waitingForDependentsDeletion":null}

ncdc commented:
/retest

pohly reviewed (commented):
This now looks good to me.

It's worth noting that the existing log messages alternate between "object" and "item". "item" seems to be a bit more common, so it makes sense to me that this is now used consistently.

Can you squash your commits into one? You can leave mine separate, it is fine on its own. When that is done, I'll add the LGTM.

ncdc commented:
@pohly thanks! Squashed!

ncdc commented:
/test pull-kubernetes-unit

pohly reviewed (approved):
/lgtm
/assign @soltysh 

For approval.

soltysh commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -287,20 +289,21 @@ func (gb *GraphBuilder) IsSynced() bool {
 	// Set up the stop channel.
 	gb.monitorLock.Lock()
-	gb.stopCh = stopCh
+	gb.stopCh = ctx.Done()
```
Nit: we could probably get rid of the `stopCh` entirely and replace it with context where needed, especially that I only found one place as such :wink:

soltysh reviewed (approved):
/approve

soltysh commented:
/priority important-longterm
